### PR TITLE
Change host associations so they are not "promoted" to !fir.ptr

### PR DESCRIPTION
**Pull request by schweitzpgi**
(POINTER) variables. The promotion negatively impacts alias analysis and
performance.

In the initial implementation of the feature, host associated variables
were always passed as if they had been declared with the POINTER
attribute even when they had not been. These changes allow non-pointer,
non-allocatable variables to be passed as !fir.ref or !fir.box and
maintain the non-aliasing properties of the original host associated
variable whenever appropriate.

This change also allows descriptors to be created which are not
interoperatble with BIND(C) but do allow a variable to be boxed with its
properties. These properties can then be passed to internal procedures
without loss of information from the host. An attribute can be used to
force a descriptor to be created in a normalized form compatible with
the BIND(C) requirements from the standard.

# **DO NOT FILE A PULL REQUEST**

We are in the process of a github migration.  Please do not create a pull request as this could interfere with the process.

### CODE REVIEW

**Review by psteinfeld — APPROVED** [review]
All builds, tests, and looks good to me.

You might want to get a review from someone more familiar with the code before merging, though.

**Review by jeanPerier — COMMENTED** [review]
Appart from the comment on the lower bounds, I was thinking if it would always be OK to get rid of the pointer attribute, and there is one case where I think this might currently lead to trouble: equivalence from the host:

```
subroutine foo()
   integer :: i(4) = [1,2,3,4]
   integer :: j(4)
   equivalence (i(1), j(1))
   call bar()
   print *, i
   contains
     subroutine bar()
       i = j(4:1:-1)
     end subroutine
end subroutine
 call foo()
end
```

I think we would now miss i/j overlap in bar.

Otherwise, the change looks good to me, thanks.

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp
Comment (schweitzpgi) [reply flang/lib/Optimizer/CodeGen/CodeGen.cpp:1496]: EQUIVALENCEd variables remain broken with this patch. They still need to be addressed in a different way than conservatively making copies everywhere.

**Pull request merged**
